url6a9f553762_Bresiu_socket_io_tgz-pJ8-MainJ8/src/com/github/nkzawa/engineio/client/Socket.java

```diff
@@ -0,0 +1,779 @@
+                    Socket.this.transport.close();
+                }
+
+            }
+        });
+        return this;
+    }
+
+    private void onError(Exception err) {
+        logger.fine(String.format("socket error %s", err));
+        Socket.priorWebsocketSuccess = false;
+        this.emit(EVENT_ERROR, err);
+        this.onClose("transport error", err);
+    }
+
+    private void onClose(String reason) {
+        this.onClose(reason, null);
+    }
+
+    private void onClose(String reason, Exception desc) {
+        if (this.readyState == ReadyState.OPENING || this.readyState == ReadyState.OPEN) {
+            logger.fine(String.format("socket close with reason: %s", reason));
+            final Socket self = this;
+
+            // clear timers
+            if (this.pingIntervalTimer != null) {
+                this.pingIntervalTimer.cancel(true);
+            }
+            if (this.pingTimeoutTimer != null) {
+                this.pingTimeoutTimer.cancel(true);
+            }
+
+            EventThread.nextTick(new Runnable() {
+                @Override
+                public void run() {
+                    self.writeBuffer.clear();
+                    self.callbackBuffer.clear();
+                    self.prevBufferLen = 0;
+                }
+            });
+
+            // stop event from firing again for transport
+            this.transport.off(EVENT_CLOSE);
+
+            // ensure transport won't stay open
+            this.transport.close();
+
+            // ignore further transport communication
+            this.transport.off();
+
+            // set ready state
+            this.readyState = ReadyState.CLOSED;
+
+            // clear session id
+            this.id = null;
+
+            // emit close events
+            this.emit(EVENT_CLOSE, reason, desc);
+        }
+    }
+
+    /*package*/ List<String > filterUpgrades(List<String> upgrades) {
+        List<String> filteredUpgrades = new ArrayList<String>();
+        for (String upgrade : upgrades) {
+            if (this.transports.contains(upgrade)) {
+                filteredUpgrades.add(upgrade);
+            }
+        }
+        return filteredUpgrades;
+    }
+
+    public static class Options extends Transport.Options {
+
+        /**
+         * List of transport names.
+         */
+        public String[] transports;
+
+        /**
+         * Whether to upgrade the transport. Defaults to `true`.
+         */
+        public boolean upgrade = true;
+
+        public boolean rememberUpgrade;
+        public String host;
+        public String query;
+
+
+        private static Options fromURI(URI uri, Options opts) {
+            if (opts == null) {
+                opts = new Options();
+            }
+
+            opts.host = uri.getHost();
+            opts.secure = "https".equals(uri.getScheme()) || "wss".equals(uri.getScheme());
+            opts.port = uri.getPort();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

`getPort()` returns -1 if the port is not set. Check the value returned by `getPort`. Use a default port if the value returned is negative. See the following diffs on GitHub that make this code change: [diff1](https://github.com/apache/skywalking/commit/10d55860dbb4a4c6fec20d54c1a61382bc681481#diff-d4c3e5510bbb6120dc300aa7b2b1c9051d39fd70662af287db112833c7944e11L43), [diff2](https://github.com/opennetworkinglab/onos/commit/6381a24e9d48d4d4e7110cf78b91a6ab47f791d0#diff-7678808b29b9776d35b4ec9d00428db2fa2e2adcfc6ffc7a69812f3ab330fe10L114)

Learn more: [URL.getPort in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/net/URL.html#getPort--)